feat(Table): table settings component added

TaPo4eK111291 opened the pull request.

beliarh commented on src/components/TableSettingsColumn/TableSettingsColumn.tsx:
```diff
@@ -0,0 +1,97 @@
+        id: column.id,
+    });
+
+    const style = {
+        transform: CSS.Transform.toString(transform),
+        transition,
+    };
```
There must be a bug with `transform` style:

https://github.com/user-attachments/assets/db194aa0-3cfa-4b17-8d06-71096bb8d2bf

TaPo4eK111291 replied:
Changed layout and set override for scaleY to 1, now it's looks much better, thx.

beliarh reviewed (changes requested):
1. Is it possible to prevent some columns from hiding? (by disabling checkbox for example)
2. Parent column's checkbox is not clickable for some reason
3. What about supporting the column pinning feature?

TaPo4eK111291 commented on src/utils/addComponentKeysets.ts:
```diff
@@ -0,0 +1,12 @@
+
+import type {Lang} from './configure';
+
+export function addComponentKeysets<T extends KeysData>(data: Record<Lang, T>, keysetName: string) {
```
Reused addComponentsKeysets from @gravity-ui/uikit/i18n

TaPo4eK111291 commented on src/components/TableSettingsColumn/TableSettingsColumn.tsx:
```diff
@@ -0,0 +1,97 @@
+    return (
+        <div
+            key={column.id}
+            className={b('column')}
```
Don't want to mix styles of table-settings with table-settings-column, but I left the isolated component and corrected the classes to:
gt-table-settings-column__column => gt-table-settings-column
gt-table-settings-column__column-content => gt-table-settings-column__content

TaPo4eK111291 commented on src/components/TableSettings/TableSettings.tsx:
```diff
@@ -0,0 +1,110 @@
+
+    return (
+        <React.Fragment>
+            <Popup open={open} onClose={() => setOpen(false)} anchorRef={anchorRef}>
```
Set position like uikit table.

TaPo4eK111291 commented on package.json:
```diff
@@ -43,6 +43,7 @@
   },
   "dependencies": {
     "@bem-react/classname": "^1.6.0",
+    "@gravity-ui/i18n": "^1.6.0",
```
Moved

TaPo4eK111291 commented on src/components/TableSettings/TableSettings.tsx:
```diff
@@ -0,0 +1,110 @@
+    const [open, setOpen] = React.useState<boolean>(false);
+    const columns = table.getAllColumns();
+    const filteredColumns = column
+        ? columns.filter((otherColumn) => otherColumn.id !== column.id)
```
Wrapped computed properties by useMemo.

TaPo4eK111291 commented on src/components/TableSettings/__stories__/stories/TableSettingsColumnStory.tsx:
```diff
@@ -0,0 +1,103 @@
+    contactPriority: string;
+};
+
+const data = [
```
Added

TaPo4eK111291 commented on src/components/TableSettings/__stories__/stories/TableSettingsColumnStory.tsx:
```diff
@@ -0,0 +1,103 @@
+    {
+        id: 'contact_group_group',
+        header: 'Contact Group',
+        size: 100,
```
Looks like, parent column size ignores by Table, if all children has sizes. Removed sized from parent columns.

TaPo4eK111291 commented:
> * Is it possible to prevent some columns from hiding? (by disabling checkbox for example)
> * Parent column's checkbox is not clickable for some reason
> * What about supporting the column pinning feature?

1. Added reaction to enableHiding prop in columns config. Now u can set it to false, for prevent toggling.
2. Fixed.
3. Right now we don't have mockups for pinning and disabling other behaviours, we'll add them separately.

Raubzeug commented:
Name of group name should be bold.
<img width="193" alt="Screenshot 2024-09-24 at 11 36 58" src="https://github.com/user-attachments/assets/567aaf5c-3084-46c9-8fe8-cc6aa3c5efdc">

Raubzeug commented:
Drag handler is not centered:
<img width="175" alt="Screenshot 2024-09-24 at 11 38 09" src="https://github.com/user-attachments/assets/e224949a-6a8c-4037-a6a5-b6120e482bd3">

Raubzeug commented:
Column with settings handler should always be visible (pinned on the end of the flow (depends on rtl or ltr)).

TaPo4eK111291 commented:
> Name of group name should be bold. <img alt="Screenshot 2024-09-24 at 11 36 58" width="193" src="https://private-user-images.githubusercontent.com/42094364/370203236-567aaf5c-3084-46c9-8fe8-cc6aa3c5efdc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcxNzQwNTcsIm5iZiI6MTcyNzE3Mzc1NywicGF0aCI6Ii80MjA5NDM2NC8zNzAyMDMyMzYtNTY3YWFmNWMtMzA4NC00NmM5LThmZTgtY2M2YWEzYzVlZmRjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI0VDEwMjkxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBlYThmNWU1MjllMjc5Yjg5YjQ2N2IzYTNlN2VmYzEyNGQzZjM3MzQ0MjU4MWIwMDZkZWNmMzhlMzQwNjZmZjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.OTAYytRxfiroWlGhlmSUnpqJOfeL0GWQtSTurKQa9dE">

Fixed

TaPo4eK111291 commented on src/components/TableSettings/TableSettings.tsx:
```diff
@@ -0,0 +1,120 @@
+            const lastInGroup = index === renderedColumns.length - 1;
+
+            return (
+                <React.Fragment key={innerColumn.id}>
```
Removed React.Fragment

TaPo4eK111291 commented on src/components/TableSettings/TableSettings.scss:
```diff
@@ -0,0 +1,13 @@
+
+#{$block} {
+    &__popover-content {
+        padding: 4px;
```
Fixed.

TaPo4eK111291 commented on src/components/TableSettingsColumn/TableSettingsColumn.tsx:
```diff
@@ -0,0 +1,129 @@
+    };
+
+    return (
+        <div style={style} ref={setNodeRef}>
```
Fixed

TaPo4eK111291 commented on src/components/TableSettings/TableSettings.tsx:
```diff
@@ -0,0 +1,132 @@
+        });
+    };
+
+    const applyNewSettings = () => {
```
Added onSettingsApply callback for TableSettings component

TaPo4eK111291 commented on src/constants/settingsColumn.tsx:
```diff
@@ -0,0 +1,26 @@
+import {TableSettings} from '../components/TableSettings/TableSettings';
+import type {TableSettingsOptions} from '../components/TableSettings/TableSettings';
+
+export const settingsColumn: ColumnDef<unknown> = {
```
Removed

TaPo4eK111291 commented on src/components/TableSettings/TableSettings.utils.ts:
```diff
@@ -0,0 +1,80 @@
+        orderItem({id: 'root', columns: items} as Column<TData, unknown>, orderState).columns ?? [];
+
+    const handleDragEnd = ({active, over}: DragEndEvent) => {
+        const activeContainer = findContainer(orderState, active.id as string);
```
Removed all type casting

TaPo4eK111291 commented on src/components/TableSettings/TableSettings.tsx:
```diff
@@ -0,0 +1,120 @@
+        );
+    }, [headers]);
+
+    const [visibilityState, setVisibilityState] = React.useState(table.getState().columnVisibility);
```
Fixed

TaPo4eK111291 commented on src/utils/configure.ts:
```diff
@@ -0,0 +1,33 @@
+export enum Lang {
```
Removed all copypaste, thx.

beliarh commented on src/components/TableSettings/TableSettings.utils.ts:
```diff
@@ -0,0 +1,113 @@
+) => {
+    const [activeDepth, setActiveDepth] = React.useState<number | undefined>();
+
+    const dephMap = React.useMemo(() => {
```
`dephMap` => `depthMap`

TaPo4eK111291 replied:
Fixed

beliarh commented on src/components/TableSettingsColumn/TableSettingsColumn.utils.ts:
```diff
@@ -0,0 +1,13 @@
+    return visibilityState[column.id] ?? true;
+};
+
+export const isEnabledHidding = <TData extends unknown>(column: Column<TData>): boolean =>
```
`isEnabledHidding` => `isEnabledHiding`

TaPo4eK111291 replied:
Fixed
